### PR TITLE
materials: Convert to es6.

### PR DESCRIPTION
**Pull request by linbingquan**
Clean up for https://github.com/mrdoob/three.js/issues/19986#issuecomment-674046275

### CODE REVIEW

**Comment by linbingquan** [comment]
@Mugen87 This PR is not allow to change `examples/jsm`'s files like https://github.com/mrdoob/three.js/pull/19989#discussion_r471582910 . It that right?

**Comment by Mugen87** [comment]
At least not `LineMaterial`. In general, don't modify files in `examples/jsm` which have counterparts in `examples/js`.

Please keep in mind that such files are always authored in `examples/js` and the respective modules generated with the modularization script.

File: examples/jsm/lines/LineMaterial.js [hunk 1]
```diff
@@ -243,159 +243,160 @@ ShaderLib[ 'line' ] = {
...
+		this.setValues( parameters );
 
-};
+	}
 
-LineMaterial.prototype = Object.create( ShaderMaterial.prototype );
-LineMaterial.prototype.constructor = LineMaterial;
+}
 
 LineMaterial.prototype.isLineMaterial = true;
```
Comment (marcofugaro) [inline examples/jsm/lines/LineMaterial.js:401]: could you use the `Object.defineProperty()` pattern demonstrated in #20110 across the files modified in this PR? 🙏

Also support for class fields is coming soon.

Comment (linbingquan) [reply examples/jsm/lines/LineMaterial.js]: OK.

**Comment by DefinitelyMaybe** [comment]
Hey @linbingquan, for this PR, could we please stick to `src/` files for the moment? Like a couple of my PR's there's also been a number of changes to the files since this PR was created which creates all the current conflicts

File: src/materials/SpriteMaterial.js [hunk 2]
```diff
@@ -11,49 +11,50 @@ import { Color } from '../math/Color.js';
...
 
-	return this;
+		return this;
 
-};
+	}
 
+}
+
+SpriteMaterial.prototype.isSpriteMaterial = true;
```
Comment (Mugen87) [inline src/materials/SpriteMaterial.js:58]: These lines still needs to be changed, see https://github.com/mrdoob/three.js/pull/20100#discussion_r472421671.

**Review by Mugen87 — CHANGES_REQUESTED** [review]
@linbingquan This PR can't be merged in its current form. Please focus on `src` and resolve https://github.com/mrdoob/three.js/pull/20100#discussion_r577461616.

**Comment by linbingquan** [comment]
> @linbingquan This PR can't be merged in its current form. Please focus on `src` and resolve [#20100 (comment)](https://github.com/mrdoob/three.js/pull/20100#discussion_r577461616).

Thank you for your reply, I had changed it.

**Comment by Mugen87** [comment]
The PR still modifies `examples/`.

File: src/materials/LineBasicMaterial.js [hunk 3]
```diff
@@ -12,44 +12,45 @@ import { Color } from '../math/Color.js';
...
-LineBasicMaterial.prototype.constructor = LineBasicMaterial;
+	}
 
-LineBasicMaterial.prototype.isLineBasicMaterial = true;
 
-LineBasicMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		Material.prototype.copy.call( this, source );
```
Comment (Mugen87) [inline src/materials/LineBasicMaterial.js:40]: This is not the way super methods are called with ES6. It should be:
```js
super.copy( source );
```

**Comment by Mugen87** [comment]
`LineMaterial` is derived from `ShaderMaterial`. So you can't convert this class.

File: src/materials/MeshDistanceMaterial.js [hunk 4]
```diff
@@ -22,60 +22,60 @@ import { Vector3 } from '../math/Vector3.js';
...
+		this.setValues( parameters );
 
-MeshDistanceMaterial.prototype.isMeshDistanceMaterial = true;
+	}
 
-MeshDistanceMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		super.clone( source );
```
Comment (Mugen87) [inline src/materials/MeshDistanceMaterial.js:58]: Why are you using `clone()` here?!?

File: src/materials/MeshMatcapMaterial.js [hunk 5]
```diff
@@ -33,82 +33,83 @@ import { Color } from '../math/Color.js';
...
-MeshMatcapMaterial.prototype.constructor = MeshMatcapMaterial;
+	}
 
-MeshMatcapMaterial.prototype.isMeshMatcapMaterial = true;
 
-MeshMatcapMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		super.clone( source );
```
Comment (Mugen87) [inline src/materials/MeshMatcapMaterial.js:80]: `clone()`?!?

File: src/materials/MeshNormalMaterial.js [hunk 6]
```diff
@@ -28,70 +28,70 @@ import { Vector2 } from '../math/Vector2.js';
...
+		this.setValues( parameters );
 
-MeshNormalMaterial.prototype.isMeshNormalMaterial = true;
+	}
 
-MeshNormalMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		super.clone( source );
```
Comment (Mugen87) [inline src/materials/MeshNormalMaterial.js:69]: ...

File: src/materials/MeshToonMaterial.js [hunk 7]
```diff
@@ -42,104 +42,104 @@ import { Color } from '../math/Color.js';
...
+		this.setValues( parameters );
 
-MeshToonMaterial.prototype.isMeshToonMaterial = true;
+	}
 
-MeshToonMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		super.clone( source );
```
Comment (Mugen87) [inline src/materials/MeshToonMaterial.js:100]: ...

File: src/materials/PointsMaterial.js [hunk 8]
```diff
@@ -15,50 +15,50 @@ import { Color } from '../math/Color.js';
...
+		this.setValues( parameters );
 
-PointsMaterial.prototype.isPointsMaterial = true;
+	}
 
-PointsMaterial.prototype.copy = function ( source ) {
+	copy( source ) {
 
-	Material.prototype.copy.call( this, source );
+		super.clone( source );
```
Comment (Mugen87) [inline src/materials/PointsMaterial.js:45]: ...

**Comment by Mugen87** [comment]
Sorry, you should really focus more when you apply changes. The last changes were quite sloppy.

**Comment by linbingquan** [comment]
> Sorry, you should really focus more when you work on these PRs. The last changes were quite sloppy.

I am so sorry, wasted your valuable time, I make a mistake at first commit, I am confusion why this PR not pass the E2E testing, I am not strictly review by myself.

image

**Comment by linbingquan** [comment]
> `LineMaterial` is derived from . So you can't convert this class.`ShaderMaterial`

Class `LineMaterial` extends Class `ShaderMaterial` and  Class `ShaderMaterial` extends Class `Material`, other class material extends Class `Material`.

If not to change `examples/jsm/lines/LineMaterial.js`, It is hard to covert to ES6...

**Comment by Mugen87** [comment]
Well, it seems it's necessary to stall this PR until `LineMaterial` can be converted.

File: examples/jsm/lines/LineMaterial.js (hunk 1 above)
Comment (Mugen87) [reply examples/jsm/lines/LineMaterial.js]: According to #21293, we should not use `Object.defineProperty()` anymore. So e.g.
```js
LineMaterial.prototype.isLineMaterial = true;
```
was actually fine.

Sorry for this back and forth!

Comment (linbingquan) [reply examples/jsm/lines/LineMaterial.js]: > Sorry for this back and forth!

It doesn't matter. It's not a big problem.

**Comment by linbingquan** [comment]
> Or you decide to undo the changes to `LineMaterial`, `ShaderMaterial` and `Material`.

@Mugen87 Thank you for your help. I think this PR is OK now.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Comment by Mugen87** [comment]
This PR actually broke: `GLTFMeshStandardSGMaterial`

https://github.com/mrdoob/three.js/blob/a0366012826d208045e069a2858a0f3987e09dbf/examples/js/loaders/GLTFLoader.js#L1063-L1065

@linbingquan I suggest you revert the changes to `MeshStandardMaterial `.

**Comment by Mugen87** [comment]
Never mind: #21319

**Comment by mrdoob** [comment]
Good catch!

**Comment by donmccurdy** [comment]
Sorry that GLTFLoader held this back — can we change GLTFMeshStandardSGMaterial to an ES6 class as well, or is there more to it than that?

**Comment by mrdoob** [comment]
It's all good! I think we made quite a lot of progress this month in the transition regardless.

**Pull request merged**